### ISSUE TITLE
How is auto reload supposed to work?

### ISSUE THREAD

**Issue by ericmorand** [twigphp/Twig#2697]
Hi guys.

We faced a bug recently with Twing where the auto_reload option was not working as expected:

https://github.com/ericmorand/twing/issues/194

It was fixed by changing the loadTemplate method of our TwingEnvironment class, which is analoguous to TwigPHP Twig_Environment one.

Made me wonder how auto_reload can be working in TwigPHP:

```
public function loadTemplate($name, $index = null)
    {
        $cls = $mainCls = $this->getTemplateClass($name);
        if (null !== $index) {
            $cls .= '_'.$index;
        }

        if (isset($this->loadedTemplates[$cls])) {
            return $this->loadedTemplates[$cls];
        }
        
        if (!class_exists($cls, false)) {
            $key = $this->cache->generateKey($name, $mainCls);

            if (!$this->isAutoReload() || $this->isTemplateFresh($name, $this->cache->getTimestamp($key))) {
                $this->cache->load($key);
            }
            // ...
    }
```

When I read the code, I can't unsee that if the template has already been loaded previously, it is returned as is and not be given the chance to be reloaded.

Is there something I'm missing there? In Twing, we had to put the `loadedTemplates` check inside the `isAutoReload `check. to fix the bug. But I'm not sure it is the correct way to fix that, maybe the bug is coming from somewhere else.

What do you think?

**Comment by jvasseur** [comment]
The template can't be reloaded inside the same request because it would need to redeclare the corresponding PHP class which is impossible.

The "auto-reload" feature means the template will be reloaded in subsequent requests if the source changed, but we can't do anything in the same request because of how Twig works.

**Comment by ericmorand** [comment]
@jvasseur, oh, so on each subsequent request, a fresh new instance of Twig_Environment is created and thus loadedTemplates is empty. Only the external cache is potentially still valid.

**Comment by jvasseur** [comment]
In PHP everything is destroyed between request so yes, each request will need to build a new `Twig_Environment` instance, thus getting a empty `loadedTemplates` array.

**Comment by ericmorand** [comment]
Thanks a lot, that makes a lot of sense. I think I'll keep the same logic with Twing, for consistence sake with TwigPHP. Middlewares will have to create a fresh instance of the environment on each build trip.

**Comment by stof** [comment]
@ericmorand be careful. In Node, you run a single process, not a process per request. So the environment would not be recreated for each request. You might want to make auto_reload work differently there (based on JS rules rather than PHP ones)

**Comment by ericmorand** [comment]
@stof, you're right. I had to change the way auto_reload works to fix the bug.

But I then discovered that the "one fresh instance per request - or per build round-trip if you think about a Node process that would watch files" is the best way to ensure a consistent and reliable behavior.

For example, there is no easy way to catch changes on extended templates without restarting with a fresh instance. When a template has a parent, this parent is loaded in the *constructor* of the child template. It means that, with auto_reload set to true, if the parent template change, it is never given the chance to be reloaded since loadTemplate is never called on him. It may be called on the child template itself but since the child template didn't change (only its parent changed), it's resolved from the local environment cache (loadedTemplates) which contains already instanciated templates. Thus the template constructor is not called.

I don't want to mess with TwigPHP logic. First because it works perfectly, is realiable and has been tested and improved over the years. Second because I want to stick with my principle of being the closest possible to your implementation.

That's why I reverted the fix. And everything is working wonderfully as long as we create a fresh Twing instance before each subsequent build.
